### ISSUE TITLE
Casting error while building SRep with SlicerSALT

### ISSUE THREAD

**Issue by lopezmt** [KitwareMedical/SlicerSkeletalRepresentation#20]
Hi,

I have been trying to build SlicerSALT using the qt5-centos7 Slicer build environment Docker file but the compilation was ending with errors when trying to build SRep.

Those are the steps to reproduce:
After creating a container with the slicer/buildenv-qt5-centos7:latest image I used the following commands:

```
mkdir SlicerSALT
cd SlicerSALT/
git clone https://github.com/kitware/SlicerSALT
mkdir SlicerSALT-build
cd SlicerSALT-build
cmake ../SlicerSALT-build -DSlicer_REQUIRED_QT_VERSION=5.11.2
make -j4
```

The compilation was ending with this error:
```
[ 74%] Building CXX object E/SRep/SkeletalRepresentationInitializer/Logic/CMakeFiles/vtkSlicerSkeletalRepresentationInitializerModuleLogicPythonD.dir/vtkBackwardFlowLogicPython.cxx.o

/osm_qSlicerUnitsModuleWidgets_module_init.cpp.o
/NIRAL/work/lpzmateo/source/Projects/SlicerSALT-build/Slicer-build/E/SRep/SkeletalRepresentationInitializer/Logic/vtkBackwardFlowLogicPython.cxx: In function ‘PyObject* PyvtkBackwardFlowLogic_runApplyTPS(PyObject*, PyObject*)’:
/NIRAL/work/lpzmateo/source/Projects/SlicerSALT-build/Slicer-build/E/SRep/SkeletalRepresentationInitializer/Logic/vtkBackwardFlowLogicPython.cxx:25:68: error: invalid static_cast from type ‘vtkObjectBase*’ to type ‘vtkBackwardFlowLogic*’
   vtkBackwardFlowLogic *op = static_cast<vtkBackwardFlowLogic *>(vp);
                                                                    ^
/NIRAL/work/lpzmateo/source/Projects/SlicerSALT-build/Slicer-build/E/SRep/SkeletalRepresentationInitializer/Logic/vtkBackwardFlowLogicPython.cxx: In function ‘PyObject* PyvtkBackwardFlowLogic_computePairwiseTPS(PyObject*, PyObject*)’:
/NIRAL/work/lpzmateo/source/Projects/SlicerSALT-build/Slicer-build/E/SRep/SkeletalRepresentationInitializer/Logic/vtkBackwardFlowLogicPython.cxx:55:68: error: invalid static_cast from type ‘vtkObjectBase*’ to type ‘vtkBackwardFlowLogic*’
   vtkBackwardFlowLogic *op = static_cast<vtkBackwardFlowLogic *>(vp);
                                                                    ^
/NIRAL/work/lpzmateo/source/Projects/SlicerSALT-build/Slicer-build/E/SRep/SkeletalRepresentationInitializer/Logic/vtkBackwardFlowLogicPython.cxx: In function ‘PyObject* PyvtkBackwardFlowLogic_generateEllipsoidSrep(PyObject*, PyObject*)’:
/NIRAL/work/lpzmateo/source/Projects/SlicerSALT-build/Slicer-build/E/SRep/SkeletalRepresentationInitializer/Logic/vtkBackwardFlowLogicPython.cxx:91:68: error: invalid static_cast from type ‘vtkObjectBase*’ to type ‘vtkBackwardFlowLogic*’
   vtkBackwardFlowLogic *op = static_cast<vtkBackwardFlowLogic *>(vp);
                                                                    ^
make[5]: *** [E/SRep/SkeletalRepresentationInitializer/Logic/CMakeFiles/vtkSlicerSkeletalRepresentationInitializerModuleLogicPythonD.dir/vtkBackwardFlowLogicPython.cxx.o] Error 1
make[4]: *** [E/SRep/SkeletalRepresentationInitializer/Logic/CMakeFiles/vtkSlicerSkeletalRepresentationInitializerModuleLogicPythonD.dir/all] Error 2
make[4]: *** Waiting for unfinished jobs....
make[3]: *** [all] Error 2
make[2]: *** [slicersources-build/Slicer-prefix/src/Slicer-stamp/Slicer-build] Error 2
make[1]: *** [slicersources-build/CMakeFiles/Slicer.dir/all] Error 2
make: *** [all] Error 2
```

To solve this issue we modified, in the header file vtkBackwardFlowLogic.h, the declaration of the class vtkBackwardFlowLogic so it inherits from vtkObjectBase. It solved the casting issue.

Do you think this was the good way to solve the issue?

Thanks,
Mateo

**Comment by phcerdan** [comment]
Thanks @lopezmt for the detailed report.
I think makes all the sense, especially when the class is named vtkXXX..... Not sure why we were not hitting this before.
 
Could you please open a PR with the fix? We want the same fix for vtkForwardFlowLogic.

**Comment by phcerdan** [comment]
@ZhiyLiu have a look at this issue for the future. You mentioned in #19 that a refactor of vtkBackwardFlowLogic was on the way.

**Comment by lopezmt** [comment]
@phcerdan, no problem I will  create the PR

**Comment by phcerdan** [comment]
closed in #22
